### PR TITLE
Add DTH Plugin Registration recovery.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Also modified all plugin activities to no longer throw but to send down error messages
- Separated message handling and message processing.
- Added faulted registration recovery bits.

#1405

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@NTaylorMullen__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Taylor Mullen). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla2.msopentech.com.
    </p>

TTYL, MSOTBOT;

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 1]
```diff
@@ -490,6 +497,34 @@ private bool PerformCompilation()
...
         }
 
+        private bool PerformPluginWork()
+        {
+            if (_pluginRegistration.WasAssigned ||
+                _pluginWorkNeeded.WasAssigned ||
+                _filesChanged.WasAssigned ||
+                _rebuild.WasAssigned ||
+                _restoreComplete.WasAssigned ||
+                _sourceTextChanged.WasAssigned)
```
Comment (davidfowl) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Do you need to actually do something for all of these types of messages? That doesn't seem right. You only need to do work if there is a plugin registered.

Comment (NTaylorMullen) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: I can expose a `WorkNeeded` property on the `PluginHandler` to short-circuit all of this. I just figured the extra checking here was pretty quick and the `Process` method basically no-ops if there's nothing to do.

Comment (davidfowl) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: I'm not worried about the verbosity of the if statement. I'm not sure this is correct. No work should happen if a plugin is never registered.

**Comment by NTaylorMullen** [comment]
Updated.

File: src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs [hunk 2]
```diff
@@ -490,6 +501,37 @@ private bool PerformCompilation()
...
+        private bool ShouldPerformPluginWork()
+        {
+            return 
+                _pluginRegistration.WasAssigned ||
+                _pluginWorkNeeded.WasAssigned ||
+                (_pluginHandler.FaultedRegistrations &&
+                (_filesChanged.WasAssigned ||
+                _rebuild.WasAssigned ||
+                _restoreComplete.WasAssigned ||
+                _sourceTextChanged.WasAssigned));
```
Comment (davidfowl) [inline src/Microsoft.Framework.DesignTimeHost/ApplicationContext.cs]: Get rid of this one.

**Comment by NTaylorMullen** [comment]
Updated.

**Comment by NTaylorMullen** [comment]
Updated. Did not make `IAssemblyLoadContext` dynamic per `IPlugin`.

**Comment by NTaylorMullen** [comment]
Good?

**Comment by NTaylorMullen** [comment]
Getting this in tonight unless there's an objection :smile:

**Pull request merged**